platform/linux-generic/odp_ml.c

```diff
@@ -0,0 +1,2673 @@
+
+	int retval		= -1; /* Return value of this function */
+	int ret			= 0;
+	OrtStatus *status	= NULL;
+	uint32_t batch_size	= 0;
+
+	OrtValue *input_tensor[CONFIG_ML_MAX_INPUTS]	= {0};
+	OrtValue *output_tensors[CONFIG_ML_MAX_OUTPUTS]	= {0};
+	const char *input_names[CONFIG_ML_MAX_INPUTS]	= {0};
+	const char *output_names[CONFIG_ML_MAX_OUTPUTS]	= {0};
+
+	const OrtApi *ort_api			= _odp_ml_glb->ort_api;
+	ml_model_t *mdl				= ml_model_from_handle(model);
+	const odp_ml_model_info_t *ml_info	= &mdl->info;
+	const odp_ml_input_info_t *input_info	= mdl->input_info;
+	const odp_ml_output_info_t *output_info = mdl->output_info;
+	OrtSession *session			= mdl->session;
+
+	if (!(_odp_ml_glb->ml_config.run_mode_mask & ODP_ML_COMPL_MODE_SYNC)) {
+		_ODP_ERR("Synchronous running is not enabled\n");
+		return -1;
+	}
+
+	odp_ticketlock_lock(&mdl->lock);
+	if (odp_unlikely(mdl->state != ML_STATE_LOADED)) {
+		_ODP_ERR("Wrong model state: not created or not loaded\n");
+		odp_ticketlock_unlock(&mdl->lock);
+		return -1;
+	}
+	mdl->state = ML_STATE_INFERENCING;
+	odp_ticketlock_unlock(&mdl->lock);
+
+	memset(&result_local, 0, sizeof(result_local));
+
+	if (ODP_DEBUG && verify_run_params(model, data, param)) {
+		result_local.error_code = ML_BAD_INPUT;
+		goto init_fail;
+	}
+
+	if (param && param->batch_size)
+		batch_size = param->batch_size;
+
+	uint32_t seg_idx = 0;
+
+	/* Transfer input data to tensor */
+	for (uint32_t i = 0; i < ml_info->num_inputs; i++) {
+		ret = input_data_to_tensor(&input_info[i],
+					   data->num_input_seg,
+					   data->input_seg,
+					   &seg_idx,
+					   batch_size,
+					   &input_tensor[i]);
+		if (ret) {
+			_ODP_ERR("%uth input data to tensor failed\n", i);
+			result_local.error_code = ML_LIB_FAILED;
+			goto release_input_tensors;
+		}
+
+		_ODP_DBG("input_tensor[%u]: %p\n", i, input_tensor[i]);
+
+		/* Model input names */
+		input_names[i] = input_info[i].name;
+	}
+
+	if (seg_idx < data->num_input_seg) {
+		_ODP_ERR("Excess input segments\n");
+		ret = -1;
+	}
+
+	for (uint32_t i = 0; i < ml_info->num_outputs; i++)
+		output_names[i] = output_info[i].name;
+
+	/* Run inference */
+	status = ort_api->Run(session,
+			      NULL,
+			      (const char * const *)input_names,
+			      (const OrtValue * const*)input_tensor,
+			      ml_info->num_inputs,
+			      (const char * const *)output_names,
+			      ml_info->num_outputs,
+			      output_tensors);
+
+	if (check_ortstatus(status)) {
+		_ODP_ERR("Run inference failed\n");
+		result_local.error_code = ML_LIB_FAILED;
+		goto release_all_tensors;
+	}
+
+	/* Verify output tensors and store them to output */
+	if (output_tensors_to_data(output_tensors, ml_info->num_outputs, param,
+				   output_info, data, &result_local)) {
+		_ODP_ERR("Output tensors to data failed\n");
+		goto release_all_tensors;
+	}
+
+	retval = 1;
+
+release_all_tensors:
+	for (uint32_t i = 0; i < ml_info->num_outputs; i++)
+		ort_api->ReleaseValue(output_tensors[i]);
```

Comment:
Is it safe to call `ReleaseValue` in cases where `num_outputs`/`num_inputs` weren't reached?

Reply:
In terms of implementation, yes. ReleaseValue() is just a delete, which is safe since the pointers here are initialized to zero. ORT API doesn't document whether ReleaseValue() may be called with NULL, but I think it's a reasonable assumption.